denoise_tests.py

```diff
@@ -0,0 +1,385 @@
+        raise ValueError
+    return result
+
+
+# OpenCV Block-matching and 3D filtering
+# Only available in non-free extension
+# I was unable to compile it so this remains untested
+def cv_bm3d(data):
+    result = np.zeros_like(data)
+    if data.ndim == 2:
+        cv.xphoto.bm3dDenoising(data, result)
+    elif data.ndim == 3:
+        for i in range(data.shape[0]):
+            cv.xphoto.bm3dDenoising(data[i], result[i])
+    else:
+        raise ValueError
+    return result
+
+
+# OpenCV non-local means denoising
+# Much faster than the scikit-image version
+def cv_nl_means(data, h):
+    norm_type = cv.NORM_L1  # required for uint16
+    if data.ndim == 2:
+        return cv.fastNlMeansDenoising(data, h=h, normType=norm_type)
+    elif data.ndim == 3:
+        slices = []
+        for s in data:
+            slices.append(cv.fastNlMeansDenoising(s, h=h, normType=norm_type))
+        return np.array(slices)
+    else:
+        raise ValueError
+
+
+# ImageJ N-D median filter
+RectangleShape = scyjava.jimport("net.imglib2.algorithm.neighborhood.RectangleShape")
+def ij_median(data, shape):
+    src = IMAGEJ.op().transform().flatIterableView(IMAGEJ.py.to_java(data))
+    dst = IMAGEJ.op().run('create.img', src)
+    return IMAGEJ.py.from_java(IMAGEJ.op().filter().median(dst, src, shape))
+
+
+# ImageJ non-local means denoising
+def ij_nl_means(data):
+    """IJ.py.to_java() returns a Java view of the numpy array.
+    It does not allocate new memory, so make a copy to avoid overwriting the input image."""
+    data_copy = data.copy()
+    plugin = "Non-local Means Denoising"
+    args = {
+        'sigma': 5,
+        'smoothing_factor': 1,
+        'auto_estimate_sigma': True
+    }
+    if data_copy.ndim == 3:
+        slices = []
+        for s in data_copy:
+            result = run_ij_plugin(s, plugin, args)
+            slices.append(result)
+        return np.array(slices)
+    elif data_copy.ndim == 2:
+        return run_ij_plugin(data_copy, plugin, args)
+    else:
+        raise ValueError
+
+
+# Scikit-image non-local means denoising
+# Much slower than OpenCV and ImageJ implementations
+def skimage_nl_means(data):
+    """This is way too slow"""
+    sigma = np.mean(estimate_sigma(data))
+    denoised = denoise_nl_means(data, h=0.8 * sigma, sigma=sigma, preserve_range=True)
+    # cast back to uint16
+    return denoised
+
+
+def denoise_fft(data):
+    """This does not work well, ringing artifacts in the result.
+    Taken from https://scipy-lectures.org/intro/scipy/auto_examples/solutions/plot_fft_image_denoise.html"""
+    im_fft = fftpack.fftn(data)
+    # Define the fraction of coefficients (in each direction) we keep
+    keep_fraction = 0.3
+    # Set s, r and c to be the number of slices, rows and columns of the array.
+    s, r, c = im_fft.shape
+    # Set to zero all slices with indices between s*keep_fraction and
+    # s*(1-keep_fraction):
+    im_fft[int(s * keep_fraction):int(s * (1 - keep_fraction)), :, :] = 0
```

Comment:
Can't quite tell, but it looks like this is a band-pass filter truncating both high and low frequencies. Is that what you meant?  I would figure we'd want to e.g. trim out speckles, which are in the high band.

Reply:
Or, maybe it's the opposite?  Just keeping the high and low frequencies?  I would expect this to look weird.

Reply:
I think it might be because of symmetry in the frequency domain image? If I do:
```python
    keep_fraction = 0.95
    im_fft[int(s * keep_fraction):, :, :] = 0
    # Similarly with the rows and columns:
    im_fft[:, int(r * keep_fraction):, :] = 0
    im_fft[:, :, int(c * keep_fraction):] = 0
```
The result is even more strange looking. Tbh I don't have a deep understanding on Fourier analysis...it looked like something to try though :D